### PR TITLE
Allow options in config from both `pip-compile` and `pip-sync`

### LINKED ISSUE #1932: `pip-compile` and `pip-sync` have different options, but single config section
#### Environment Versions

1. OS Type: any
1. Python version: any
1. pip version: any
1. pip-tools version: `pip-compile, version 7.1.0`

#### Steps to replicate

```bash
echo pytz > requirements.in
echo -e "[tool.pip-tools]\ngenerate-hashes = true" > .pip-tools.toml
pip-compile
pip-sync
```

#### Expected result

Successful sync.

#### Actual result

```console
$ pip-compile
#
# This file is autogenerated by pip-compile with Python 3.11
# by the following command:
#
#    pip-compile --generate-hashes
#
pytz==2023.3 \
    --hash=sha256:1d8ce29db189191fb55338ee6d0387d82ab59f3d00eac103412d64e0ebd0c588 \
    --hash=sha256:a151b3abb88eda1d4e34a9814df37de2a80e301e68ba0fd856fb9b46bfbbbffb
    # via -r requirements.in


$ pip-sync
Usage: pip-sync [OPTIONS] [SRC_FILES]...
Try 'pip-sync -h' for help.

Error: No such config key 'generate_hashes'.
```

### PR DESCRIPTION
**Pull request by atugushev**
Fixes #1932.

##### Contributor checklist

- [x] Provided the tests for the changes.
- [x] Assure PR title is short, clear, and good to be included in the user-oriented changelog

##### Maintainer checklist

- [x] Assure one of these labels is present: `backwards incompatible`, `feature`, `enhancement`, `deprecation`, `bug`, `dependency`, `docs` or `skip-changelog` as they determine changelog listing.
- [ ] Assign the PR to an existing or new milestone for the target version (following [Semantic Versioning](https://blog.versioneye.com/2014/01/16/semantic-versioning/)).

### CODE REVIEW

File: piptools/utils.py [hunk 1]
```diff
@@ -588,33 +588,44 @@ def _validate_config(
     :raises click.NoSuchOption: if config contains unknown keys.
     :raises click.BadOptionUsage: if config contains invalid values.
     """
-    cli_params = {
-        param.name: param
-        for param in click_context.command.params
-        if param.name is not None
+    from piptools.scripts.compile import cli as compile_cli
+    from piptools.scripts.sync import cli as sync_cli
```
Comment (chrysle) [inline piptools/utils.py:592]: Maybe move this to the top?

Comment (atugushev) [reply piptools/utils.py:592]: This avoids circular imports.

File: piptools/utils.py [hunk 2]
```diff
@@ -591,33 +591,44 @@ def _validate_config(
...
-            raise click.BadOptionUsage(
-                option_name=key,
-                message=f"Invalid value for config key {key!r}: {value!r}.",
-                ctx=click_context,
-            ) from e
+        # Validate values for both compile and sync
+        for cli_params in (compile_cli_params, sync_cli_params):
+            param = cli_params.get(key)
+            if param is None:
+                continue
```
Comment (chrysle) [inline piptools/utils.py:622]: Do we need this? I think wrong keys are caught above?

Comment (atugushev) [reply piptools/utils.py:622]: yeah, but we still checking both cli_params, see 618 line.

Comment (chrysle) [reply piptools/utils.py:622]: Sorry if I understood you wrong, but isn't this covered by the `all_keys` variable test? Also, the value can't possibly be `None`?

Comment (atugushev) [reply piptools/utils.py:622]: Suppose there is an `ask = true` option in the config. It's included in `all_keys`, but not found within  `compile_cli_params`. This Is why we skip it and proceed to validate it on `sync_cli_params`.

The condition `if key not in all_keys` is True if there is a unknown key that neither `pip-compile` nor `pip-sync` has.

Comment (chrysle) [reply piptools/utils.py:622]: Thank you for the explanantion! I actually thought both dicts were combined into one.

File: piptools/utils.py [hunk 3]
```diff
@@ -591,36 +591,47 @@ def _validate_config(
...
+
+            try:
+                param.type.convert(value=value, param=param, ctx=click_context)
```
Comment (chrysle) [inline piptools/utils.py:625]: I think you still need to update that after the merge?

```suggestion
                param.type_cast_value(value=value, ctx=click_context)
```

Comment (atugushev) [reply piptools/utils.py:625]: Yeah, I messed up with resolving conflicts :/

File: piptools/utils.py (hunk 2 above)
Comment (atugushev) [reply piptools/utils.py]: Considering your questions, I've refactored a bit in 5b029cd. Does it look all right?

File: piptools/utils.py (hunk 3 above)
Comment (atugushev) [reply piptools/utils.py]: Addressed in f0236945719081f9c012e052f785b28e8048d514

**Review by chrysle — APPROVED** [review]
Thanks for the work!

**Comment by atugushev** [comment]
@chrysle thanks for the review!

**Comment by AndydeCleyre** [comment]
Does this mean any flags between the two tools will have to be the same value, if the config file is used?

**Comment by atugushev** [comment]
@AndydeCleyre correct.

**Comment by chrysle** [comment]
BTW, I wonder, are separate config sections for the commands (for example `[tool.pip-compile]`) out of the question?

**Comment by atugushev** [comment]
@chrysle I wouldn't mind against that.

**Pull request merged**